// @expo/devcert/dist/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rootCACertPath = exports.rootCAKeyPath = exports.rootCADir = exports.caSelfSignConfig = exports.opensslDatabaseFilePath = exports.opensslSerialFilePath = exports.caVersionFile = exports.pathForDomain = exports.domainsDir = exports.configPath = exports.configDir = exports.isWindows = exports.isLinux = exports.isMac = exports.VALID_DOMAIN = exports.VALID_IP = void 0;
exports.withDomainSigningRequestConfig = withDomainSigningRequestConfig;
exports.withDomainCertificateConfig = withDomainCertificateConfig;
exports.getLegacyConfigDir = getLegacyConfigDir;
exports.ensureConfigDirs = ensureConfigDirs;
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const utils_1 = require("./utils");
function applicationConfigPath(name) {
    switch (process.platform)
